[Keyboard] Yowkees Keyball39

noroadsleft opened the pull request:
## Description

This is an import of the [official Yowkees Keyball39 source code](https://github.com/Yowkees/keyball/tree/ef172dd614839034429452b0f41a343fe004ad8e/qmk_firmware/keyboards/keyball/keyball39), updated to current QMK standards and to work with 0.25.x (original repository is based upon 0.22.14).

The Keyball39 is a split keyboard, featuring dual OLED screens, optional RGB underglow, and up to two trackballs (one per side), running on a pair of Pro Micro controllers.

I've been coordinating with the user who requested support for this board over Discord, and they've been testing .HEX files that I've been compiling from this branch. They are willing to do additional tests, but the time difference is an issue, so updates may be sporadic.

This branch as of right now includes a custom PMW3360 driver written by Yowkees. I have another branch using the core PMW3360 driver which compiles, but that branch is untested. I am unsure of any functional difference between the two.

The custom PMW3360 driver also includes two binary blobs - I am completely unsure what to do about those. As of now, the build *does not use either*, and the trackball works completely fine without either blob included. ([There are defines in the `keyball.h` lib file to include them, but both are currently commented out](https://github.com/noroadsleft/qmk_firmware/blob/b37b5f65fe04f7dad1ee3b992ab2719c4a1d1511/keyboards/yowkees/lib/keyball/keyball.h#L52-L57), as they are in the official source.)

A `via` keymap is included, but is also untested. (I should maybe have that happen.)

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

noroadsleft committed:
Yowkees Keyball39 initial bring-up

Imported via ZIP from the designer's GitHub directory.

  - https://github.com/Yowkees/keyball/tree/ef172dd614839034429452b0f41a343fe004ad8e/qmk_firmware/keyboards/keyball
    - `main` branch at 2024-06-03 17:47 -0700

noroadsleft committed:
Update keyball struct

Fixes a complation error in `yowkees/lib/keyball/keyball.c`:

```
Compiling: keyboards/yowkees/lib/keyball/keyball.c                                                 keyboards/yowkees/lib/keyball/keyball.c:54:24: error: initializer element is not constant
     .pressing_keys = { BL, BL, BL, BL, BL, BL, 0 },
                        ^
keyboards/yowkees/lib/keyball/keyball.c:54:24: note: (near initialization for ‘keyball.pressing_keys[0]’)
keyboards/yowkees/lib/keyball/keyball.c:54:28: error: initializer element is not constant
     .pressing_keys = { BL, BL, BL, BL, BL, BL, 0 },
                            ^
keyboards/yowkees/lib/keyball/keyball.c:54:28: note: (near initialization for ‘keyball.pressing_keys[1]’)
keyboards/yowkees/lib/keyball/keyball.c:54:32: error: initializer element is not constant
     .pressing_keys = { BL, BL, BL, BL, BL, BL, 0 },
                                ^
keyboards/yowkees/lib/keyball/keyball.c:54:32: note: (near initialization for ‘keyball.pressing_keys[2]’)
keyboards/yowkees/lib/keyball/keyball.c:54:36: error: initializer element is not constant
     .pressing_keys = { BL, BL, BL, BL, BL, BL, 0 },
                                    ^
keyboards/yowkees/lib/keyball/keyball.c:54:36: note: (near initialization for ‘keyball.pressing_keys[3]’)
keyboards/yowkees/lib/keyball/keyball.c:54:40: error: initializer element is not constant
     .pressing_keys = { BL, BL, BL, BL, BL, BL, 0 },
                                        ^
keyboards/yowkees/lib/keyball/keyball.c:54:40: note: (near initialization for ‘keyball.pressing_keys[4]’)
keyboards/yowkees/lib/keyball/keyball.c:54:44: error: initializer element is not constant
     .pressing_keys = { BL, BL, BL, BL, BL, BL, 0 },
                                            ^
keyboards/yowkees/lib/keyball/keyball.c:54:44: note: (near initialization for ‘keyball.pressing_keys[5]’)
 [ERRORS]
 |
 |
 |
gmake: *** [builddefs/common_rules.mk:373: .build/obj_yowkees_keyball39_default/lib/keyball/keyball.o] Error 1
gmake: *** Waiting for unfinished jobs....
```

noroadsleft committed:
Update `OLED_FONT_H`

Update to fix a compilation error caused by changing the "vendor"
directory on the initial bring-up. (Source repository uses
`keyboards/keyball/`, but this branch uses `keyboards/yowkees/`.)

```
Compiling: drivers/oled/oled_driver.c                                                              drivers/oled/oled_driver.c:27:21: fatal error: keyboards/keyball/lib/logofont/logofont.c: No such file or directory
compilation terminated.
 [ERRORS]
 |
 |
 |
gmake: *** [builddefs/common_rules.mk:373: .build/obj_yowkees_keyball39_default/oled_driver.o] Error 1
gmake: *** Waiting for unfinished jobs....
```

noroadsleft committed:
Format `info.json`

noroadsleft committed:
Add layout data to `info.json`

noroadsleft committed:
Convert `LAYOUT_no_ball` to data-driven

noroadsleft committed:
Convert `LAYOUT_left_ball` to data-driven

noroadsleft committed:
Convert `LAYOUT_right_ball` to data-driven

noroadsleft committed:
Convert `LAYOUT_dual_ball` to data-driven

noroadsleft committed:
Convert layout aliases to data-driven

noroadsleft committed:
Re-sort `info.json` keys

noroadsleft committed:
Convert `rules.mk` to data-driven

noroadsleft committed:
Convert `config.h` to data-driven

noroadsleft committed:
Add url and maintainer keys to `info.json`

noroadsleft committed:
Move RGBLIGHT and OLED rules to keyboard level

noroadsleft committed:
Change `SPLIT_HAND_MATRIX_GRID` detection behaviour

The default behaviour of `SPLIT_HAND_MATRIX_GRID` was changed in QMK
Firmware 0.24.0, but the initial code is based on 0.22.x, so this board
needs to have its detection behaviour swapped by defining
`SPLIT_HAND_MATRIX_GRID_LOW_IS_LEFT`.

diff --git a/keyboards/yowkees/keyball39/config.h b/keyboards/yowkees/keyball39/config.h
index 9b7f69b1c2..1a1560e44a 100644
--- a/keyboards/yowkees/keyball39/config.h
+++ b/keyboards/yowkees/keyball39/config.h
@@ -22,6 +22,7 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
 #define MATRIX_MASKED

 // Split parameters
+#define SPLIT_HAND_MATRIX_GRID_LOW_IS_LEFT
 #ifdef OLED_ENABLE
 #    define SPLIT_OLED_ENABLE
 #endif

noroadsleft committed:
Remove `.noci`

QMK doesn't use this file anymore, so it has been deleted.

noroadsleft committed:
Refactor keymaps

- remove redundant `quantum.h` includes
- use `LAYOUT_no_ball` macro directly
- apply four-space indent
- grid-align keycodes

noroadsleft committed:
Update readme

Fix image URL and `make` command.

drashna commented on keyboards/yowkees/keyball39/rules.mk:
```diff
@@ -0,0 +1,9 @@
+# Optical sensor driver for trackball.
+POINTING_DEVICE_DRIVER = custom
+SRC += drivers/pmw3360/pmw3360.c
+QUANTUM_LIB_SRC += spi_master.c # Optical sensor use SPI to communicate
```
Honestly, I would prefer that the custom code be removed and the core pmw3360 code be used.  

Though, the sroms should probably be removed, and shouldn't be necessary. 

But at the very least:
```suggestion
SRC += drivers/pmw3360/pmw3360.c
SPI_DRIVER_REQUIRED = yes
```

(custom is the default driver, and spi-driver-required handles the spi includes)

noroadsleft replied:
Question for my own education: 

What's the difference between following your Suggested Changes versus setting `POINTING_DEVICE_DRIVER = pmw3360` (which is what my referenced other branch does)?

noroadsleft replied:
> custom is the default driver

Near as I can tell, [this is actually not the case](https://github.com/qmk/qmk_firmware/blob/8253697a6389aaec73f85a29dd25247b32389502/builddefs/common_features.mk#L123-L162), and setting `POINTING_DEVICE_DRIVER = custom` is required.

Using `SPI_DRIVER_REQUIRED` seems fine.

I am still investigating switching to the core 3360 driver. My most recent attempt results in oversize firmware (approximately 31 KB), which I don't remember having as an issue on my other branch.

nobiliana replied:
Would be willing to help as well, however I use the keyball44 variant. I know how to code, however very green with contributing to github projects. please let me know!

noroadsleft committed:
Use `SPI_DRIVER_REQUIRED = yes` (per drashna)

noroadsleft committed:
Rename `info.json` to `keyboard.json`

noroadsleft committed:
Update `rules.mk` to set core PMW3360 driver

noroadsleft committed:
Remove custom Yowkees PMW3360 driver

noroadsleft committed:
Define `PMW33XX_CS_PIN`

noroadsleft committed:
Remove custom PMW3360 driver reference from keyball lib

noroadsleft committed:
Remove custom `pointing_device_*` functions from keyball lib

noroadsleft committed:
Remove `motion_to_mouse()` function from keyball lib

No longer used.

noroadsleft committed:
Redefine `pmw3360_MAXCPI` (as `KEYBALL_PMW3360_MAXCPI`)

This was previously defined in the custom driver. The
`keyball_set_cpi()` needs it (indirectly), so it has been redefined.

noroadsleft committed:
Use core `pmw33xx_set_cpi()` function in `keyball_set_cpi()`

QMK's core PMW3360 driver uses the true CPI value internally, so we have
to translate the Keyball library's value to QMK's expected range.

QMK's core driver also caps the range internally to a valid CPI value,
so we don't need to do it here.

noroadsleft committed:
Disable RGBLIGHT effects for default keymap

noroadsleft committed:
Enable PMW33XX SROM upload for default keymap

tzarc commented:
As of August 26, 2024, `qmk/qmk_firmware` is no longer accepting VIA-enabled keymaps as these have now transitioned to a repository under the VIA team's control.

As you've submitted a PR containing `via` or VIA-enabled keymap(s), this is your notice that they should be removed from this PR. You should now submit a secondary PR to the [VIA QMK Userspace](https://github.com/the-via/qmk_userspace_via) repository with your associated `via` or VIA-enabled keymaps instead.

noroadsleft committed:
Remove `via` keymap

noroadsleft commented:
The user who has been testing this code for me reports that none of the firmware files I've provided beyond 93bf88482912f4ea903f071d92bd4bc970af0032 work.

The revisions tested are:

- 93bf88482912f4ea903f071d92bd4bc970af0032 - fully working
- 3b2529ccd25781728961d0052fa7a71e1f9ec2cc - trackball doesn't work
- 41fbc685b3007636a798ceefaa55488f8101ff35 - trackball doesn't work

Sadly at this point, I'm out of ideas for making this work with the core driver.

podiki commented:
Hi everyone! Not sure if there is still interest from anyone, but I've been thinking about picking up the effort to upstream Keyball for QMK.

I have this branch for updates to Keyball so it build from current `qmk/qmk_firmware` (0.33.8 right now): https://github.com/podiki/keyball/tree/qmk-updates and this branch for using qmk's pmw3360 driver, based on the work above: https://github.com/podiki/keyball/tree/qemu-upstream. For the latter, it seems to work fine for me (Keyball61, one trackball on right side), though it feels a bit different from Keyball's code (might be how it translates the movement and/or cpi). I haven't yet moved to data-driven layouts and fixed other such warnings.

Edit: Oh, and using the qmk driver results in a firmware that is nearly 300 bytes larger. I haven't looked into the details of why, but with such limited space on the promicro, that is noticeable.

razorree commented:
keyball code has a lot of custom code sending data between halves if I remember correctly (I worked on it a year ago - when I bought Keyball61). I was able to remove most of the custom code and use QMK pmw3360 driver sucessfully.
